src/Modules/UngradedModules.jl

```diff
@@ -0,0 +1,3653 @@
+  return map(c,length(c)-i)
+end
+
+function module_in_complex(c::Hecke.ChainComplex, i::Int)
+  return obj(c,length(c)-i)
+end
+
+getindex(c::Hecke.ChainComplex, i::Int) = module_in_complex(c,i)
+
+function Base.getindex(r::Hecke.SRow, u::UnitRange)
+  R = base_ring(r)
+  s = sparse_row(R)
+  shift = 1-first(u)
+  for (p,v) = r
+    if p in u
+      push!(s.pos, p+shift)
+      push!(s.values, v)
+    end
+  end
+  return s
+end
+
+function Base.getindex(r::Hecke.SRow, R::AbstractAlgebra.Ring, u::UnitRange)
+  s = sparse_row(R)
+  shift = 1-first(u)
+  for (p,v) = r
+    if p in u
+      push!(s.pos, p+shift)
+      push!(s.values, v)
+    end
+  end
+  return s
+end
+
+function getindex(a::Hecke.SRow, b::AbstractArray{Int, 1})
+  if length(a.pos) == 0
+    return a
+  end
+  m = minimum(b)
+  b = sparse_row(parent(a.values[1]))
+  for (k,v) = a
+    if k in b
+      push!(b.pos, k-b+1)
+      push!(b.values, v)
+    end
+  end
+  return b
+end
+
+@doc Markdown.doc"""
+    sparse_row(A::MatElem)
+
+Convert `A` to a sparse row. 
+`nrows(A) == 1` must hold.
+"""
+function sparse_row(A::MatElem)
+  @assert nrows(A) == 1
+  return Hecke.sparse_matrix(A)[1]
+end
+
+@doc Markdown.doc"""
+    dense_row(r::Hecke.SRow, n::Int)
+
+Convert `r[1:n]` to a dense row, that is an AbstractAlgebra matrix.
+"""
+function dense_row(r::Hecke.SRow, n::Int)
+  R = base_ring(r)
+  A = zero_matrix(R, 1, n)
+  for i in intersect(r.pos, 1:n)
+    A[1,i] = r[i]
+  end
+  return A
+end
+
+##############################
+#should be in Singular.jl
+function Singular.intersection(a::Singular.smodule, b::Singular.smodule)
+  c = base_ring(a)
+  return Singular.Module(c, Singular.libSingular.id_Intersection(a.ptr, b.ptr, c.ptr))
+end
+
+function _reduce(a::Singular.smodule, b::Singular.smodule)
+  @assert b.isGB
+  p = Singular.libSingular.p_Reduce(a.ptr, b.ptr, base_ring(b).ptr)
+  return Singular.Module(base_ring(b), p)
+end
+
+function _reduce(a::Singular.svector, b::Singular.smodule)
+  @assert b.isGB
+  p = _reduce(Singular.Module(base_ring(b), a), b)[1]
+  return Singular.Module(base_ring(b), p)[1]
+end
+
+#TODO: tensor_product from Raul's H is broken
+
+
+######################################
+# Migrating test
+######################################
+@doc Markdown.doc"""
+    projection(F::FreeMod, indices::AbstractArray)
+
+Return the canonical projection from $F = R^I$ to $R^(\texttt{indices})$ where $\texttt{indices} \subset I$.
+"""
+function projection(F::FreeMod, indices::AbstractArray)
+  @assert all(x -> x <= ngens(F), indices)
+  @assert length(Set(indices)) == length(indices) # unique indices
+  R = base_ring(F)
+  G = FreeMod(R, length(indices))
+  return hom(F, G, [i in indices ? G[findfirst(x->x==i,indices)] : zero(G) for i=1:ngens(F)])
+end
+
+@doc Markdown.doc"""
+    preimage(H::SubQuoHom,elems::Vector{SubQuoElem{T}}, task::Symbol = :none) where {T}
+
+Return the preimage of the submodule generated by the Elements `elems` under $H$
```

Comment:
Why does this take a list of generators instead of an actual submodule? Even if there is a reason: shouldn't there also be a method which takes a submodule as second argument?